mmv1/third_party/terraform/services/storage/resource_storage_bucket.go.erb

```diff
@@ -1671,7 +1689,13 @@ func setStorageBucket(d *schema.ResourceData, config *transport_tpg.Config, res
 	if err := d.Set("custom_placement_config", flattenBucketCustomPlacementConfig(res.CustomPlacementConfig)); err != nil {
 		return fmt.Errorf("Error setting custom_placement_config: %s", err)
 	}
-
+	// Needs to hide rpo field for single-region buckets.
```

Comment:
This field is actually useful to multi-region and dual-region buckets. For single-region buckets `rpo` don't have any meaning. Since `location` field is used create single/multi/predefined-dual region buckets by giving GA or Region code(https://cloud.google.com/storage/docs/locations), At terraform side we can't verify whether location code is for single/multi/predefined-dual region. We need to rely on API to determine whether `rpo` will have any replication or not. So this condition is to remove `rpo` field from single-region bucket by checking `rpo` field in API response.

Reply:
Since Bucket location can not be updated after creation, Switching to single-region from dual or multi-region will cause force replacement of resource. Older bucket will be deleted and new bucket will be created which don't have `rpo` set. So it won't cause any trouble i guess. Does this answer your question?

Reply:
Thanks for your patience answering my questions! To confirm - for _single-region_ buckets, rpo will always be returned as empty by the API, and setting it client-side is an error. For dual-region or multi-region, it has an API-side default or you can set it. Changing `location` or `custom_placement_config.data_locations` causes recreation, which means that switching between regionalities will reset the resource state, so rpo values won't be persisted across that boundary - that is, we can ignore potential user confusion caused by O+C values persisting.

Single-region buckets:
1. If the user doesn't set an RPO for a single-region bucket, the API should return an empty string. It doesn't matter whether this gets set in state since it's empty.
2. If the user _sets_ rpo for a single-region bucket, the API will return an error. That's fine, since it's an API limitation.

Dual-region buckets:
1. If the user doesn't set an RPO, the API will return a default value.
2. If the user sets an RPO, the API will return the RPO set.
3. Changing between RPO types is safe since both are supported by dual region buckets, but could cause user confusion if they try to unset ASYNC_TURBO and expect it to revert to DEFAULT.

Multi-region buckets:
1. If the user doesn't set an RPO, the API will return a default value.
2. If the user sets an RPO, the API will return the RPO set.
3. ASYNC_TURBO can't be set due to API constraints; presumably it would return an error, which is fine. Since ASYNC_TURBO isn't supported we don't need to worry about changing between types.

Does that match your understanding of the resource?

Reply:
Yes.

Reply:
Okay, thanks for bearing with me. I've checked with other folks on the team and O+C is probably the best option here.